extensions/shared/paid-blocks/paid-block-edit.js

```diff
@@ -0,0 +1,114 @@
+
+/**
+ * WordPress dependencies
+ */
+import { Fragment, useEffect, useState } from '@wordpress/element';
+import { createHigherOrderComponent, compose } from '@wordpress/compose';
+import { BlockControls } from '@wordpress/block-editor';
+import { ToolbarGroup, Button } from '@wordpress/components';
+import { withSelect, withDispatch } from '@wordpress/data';
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import { getUpgradeUrl } from '../plan-utils';
+
+const JetpackPaidBlockEdit = OriginalBlockEdit => props => {
+	const {
+		plan,
+		postId,
+		postType,
+		planSlug,
+		isSavingPost,
+		savePost,
+		isEditedPostAutosaveable,
+		isEditedPostDirty,
+	} = props;
+
+	const [ shouldRedirectToCheckoutPage, setShouldRedirect ] = useState( false );
+
+	const checkoutUrl = getUpgradeUrl( { plan, planSlug, postId, postType } );
+
+	useEffect( () => {
+		if ( isSavingPost ) {
+			return;
+		}
+
+	    if ( ! shouldRedirectToCheckoutPage ) {
+		    return;
+	    }
+
+	    window.location.href = checkoutUrl;
+	}, [ isSavingPost, shouldRedirectToCheckoutPage, checkoutUrl ] );
+
+	const goToCheckoutPage = () => {
+		if ( ! window?.location?.href ) {
+			return;
+		}
+
+		/*
+		 * If there are not unsaved values, redirect, or
+		 * If the post is not autosaveable, redirect.
+		 */
+		if ( ! isEditedPostDirty || ! isEditedPostAutosaveable ) {
+			return window.location.href = checkoutUrl;
+		}
+
+		// Save the post before to perform redirection.
+		savePost();
+
+		// Hack to ensuring getting the saving post status.
+		setTimeout( () => setShouldRedirect( true ), 0 );
+	};
+
+	return (
+		<Fragment>
+			<BlockControls>
+				<ToolbarGroup>
+					<Button
+						aria-label={ __( 'Upgrade to Premium to use this block', 'jetpack' ) }
+						onClick={ goToCheckoutPage }
+						label={ __(
+							'Upgrade to Premium to use this block.',
+							'jetpack'
+						) }
+						showTooltip={ true }
```

Comment:
Isn't there a case that a user may need an other plan (Personal, Business etc) for a paid block? It seems that 
`requiresPaidPlan` return the Plan name https://github.com/Automattic/jetpack/blob/0fd1874f8ee31a69e08dbb38ab0a2ae7e8496671/extensions/shared/register-jetpack-block.js#L30

not sure if it is translatable though.